app/models/internal/scheduling/BaseSchedule.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2018 Dropbox, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package models.internal.scheduling;
+
+import com.arpnetworking.commons.builder.OvalBuilder;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import net.sf.oval.constraint.NotNull;
+import net.sf.oval.constraint.ValidateWithMethod;
+
+import java.time.ZonedDateTime;
+import java.util.Optional;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+/**
+ * Abstract base class for {@link Schedule}s.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+public abstract class BaseSchedule implements Schedule {
+    private final ZonedDateTime _runAtAndAfter;
+    private final Optional<ZonedDateTime> _runUntil;
+
+
+    /**
+     * Protected constructor.
+     *
+     * @param builder Instance of <code>Builder</code>.
+     */
+    protected BaseSchedule(final Builder<?, ?> builder) {
+        _runAtAndAfter = builder._runAtAndAfter;
+        _runUntil = builder._runUntil;
+    }
+
+    public ZonedDateTime getRunAtAndAfter() {
+        return _runAtAndAfter;
+    }
+
+    public Optional<ZonedDateTime> getRunUntil() {
+        return _runUntil;
+    }
+
+    @Override
+    public Optional<ZonedDateTime> nextRun(final Optional<ZonedDateTime> lastRun) {
+        Optional<ZonedDateTime> result = unboundedNextRun(lastRun);
+        while (result.isPresent() && result.get().isBefore(_runAtAndAfter)) {
+            result = unboundedNextRun(result);
+        }
+        if (_runUntil.isPresent() && result.isPresent() && result.get().isAfter(_runUntil.get())) {
+            return Optional.empty();
+        }
+        return result;
+    }
+
+    /**
+     * Return the next time the schedule should run, much as {@link Schedule}{@code .nextRun} does.
```

Comment:
nit: the comment is somewhat "public" oriented while the method should be for subclass implementation only.